Switch OVN from plaintext to TLS/SSL communication

mkalcok opened the pull request:
Changes include:
  * Generation/distribution/storage of self-signed CA
  * Automattic issuing of certificates for OVN services
  * Updates to OVN configurations to use `ssl`/`pssl` instead of `tcp`/`ptcp`
  * Automatic injection of cert paths as arguments to the `microovn.*` commands
  * New CLI subcommand to list active certificates

Note: Due to commit squashing of previous PRs, this PR looks like it contains commits that are already in the main branch.

mkalcok commented on README.md:
The README was pretty empty so I added basic installation/deployment sections along with info about self-provisioned TLS.

mkalcok commented on microovn/database/certificate.go:
This file follows the practice for other database tables where file `<subject>.go` contains `go:generate` definitions and then `<subject>.mapper.go` contains code generated based on these definitions.

fnordahl commented on microovn/ovn/certificates.go:
```diff
@@ -0,0 +1,251 @@
+
+	// Generate certificate data and properties
+	validFrom := time.Now().UTC()
+	serialNumber, err := rand.Int(rand.Reader, MaxSerialNumber)
```
The serial number must be unique per certificate per CA. As pointed out in the data structure above we should probably store this in the database and deduce the next available serial number from that information.

mkalcok replied:
In attempt to make the implementation stateless I took inspiration from [lxd](https://github.com/lxc/lxd/blob/123231d27121223f20f11a65499ce8a5e35d704b/shared/cert.go#L319). I think that with the max size of a serial being 2^128 it's quite safe to assume that we'll get unique serial numbers. On top of that I don't think that something actually breaks if there are two certificates with same serial number.

fnordahl replied:
:+1: this is fine. I remember talking to them and they brought up the requirements around serials, but if they don't care about it we probably don't need to for this version of the implementation :)

fnordahl commented on microovn/ovn/bootstrap.go:
```diff
@@ -58,12 +70,36 @@ func Bootstrap(s *state.State) error {
+	if err != nil {
+		return fmt.Errorf("failed to generate TLS certificate for ovnsb service: %s", err)
+	}
+	err = GenerateNewServiceCertificate(s, "ovn-northd", CertificateTypeClient, northdCertPath, northdKeyPath)
```
Any particular reason for using a client-certificate here?

All of the OVN daemons use the same library for DB communication. This library allows specification of any connection to be either active or passive, i.e. while unusual, northd might be told to behave like a server, expecting a connection from the DB server.

fnordahl replied:
If the KeyUsage does not contain the right bits for a server, the library will assert when attempting to create a listener, we've hit this issue in the field before. So since it is possible to configure listeners for any program based on the OVS libraries, let's just use server certs everywhere.

fnordahl commented on microovn/ovn/certificates.go:
```diff
@@ -0,0 +1,251 @@
+		validTo  time.Time
+	)
+	// Generate certificate's private key
+	keyPair, err := ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
```
What is the reasoning behind this choice of key material?

mkalcok replied:
I took again inspiration from [lxd](https://github.com/lxc/lxd/blob/123231d27121223f20f11a65499ce8a5e35d704b/shared/cert.go#L311), and I think the choice is fine. I believe that there's little difference to us whether we go with curves or RSA, aside from key sizes, as both approaches are considered secure. I don't think we need to be too concerned about whether NSA has backdoors in ECDSA curves.
Now that I think about it more, maybe we can take it down a notch and go with `p-256` instead of `p-384` for better performance and still acceptable level of security, what do you think? I honestly didn't dwell to much on this choice as the aim is not to have perfectly secure system here.

fnordahl replied:
No need to change it, I just wanted to know how we came to this conclusion. Adding a comment about why we selected this specific key material is in order, either a reference to LXD or a reference to why they ended up with that selection.

fnordahl commented on snapcraft/commands/chassis.start:
```diff
@@ -27,7 +28,12 @@ OVN_ARGS="--db-nb-addr="${OVN_LOCAL_IP}" \
+--db-sb-cluster-remote-proto=ssl \
+--ovn-controller-ssl-key="${OVN_PKIDIR}"/ovn-controller-privkey.pem \
+--ovn-controller-ssl-cert="${OVN_PKIDIR}"/ovn-controller-cert.pem \
+--ovn-controller-ssl-ca-cert="${OVN_PKIDIR}"/cacert.pem"
```
The `ovn-controller` will by default retrieve this information from the local Open vSwitch database, why not configure it there?

fnordahl replied:
The `ovn-controller` will re-read the certificate data whenever it changes on disk, so I guess this is fine.

fnordahl reviewed (changes requested):
Thank you for your work on this martin!

About the two first commits, they actually contain a white space change and subsequent revert, which is one of the reasons the previous PR was squashed. You can update your repository with `git pull --rebase` to get rid of them.

I have some comments in-line. I assume the life cycle management will appear in a subsequent PR?

mkalcok commented on microovn/database/certificate.go:
```diff
@@ -0,0 +1,33 @@
+// Certificate is used to track the Certificates used by OVN.
+type Certificate struct {
+	ID   int
+	Name string `db:"primary=yes"`
```
> The certificate contains a Serial number which uniquely identifies the certificate from this CA. Would it make sense to use that as a primary key instead?

Currently only the CA cert/key is stored in the database, certificates for ovn services are managed locally on disk. So given that the expectation is to have only one record in this table at the time, I don't think it adds too much value to store serial number as well.

> Do we need to store date of issue and expiry?

As mentioned above, non-CA certificates are not stored in the database so any tool that handles certificate lifecycle should not depend too much on the database either. I think we should rely on the database as little as possible because it'll make later integration with other certificates sources easier.

fnordahl replied:
Right, the description is misleading then. If this is only used for the CA-cert should it be named specifically for that as well?

mkalcok replied:
Oh, you mean the comment above structure, yes, that could be improved. Thanks.

fnordahl replied:
Well, not just the comment. If this database table and data structure is for use solemnly by CA Certificates, why give the impression that you are trying to implement a generic data structure for certificates?

mkalcok replied:
So, I want to preface this with the fact that I'm not supper happy about creating table that is expected to hold only single row of data. However since we need to share the CA cert and key between members I designed a generic table that can hold certificate/key pairs. Isn't it better to have generic table structure so that it's potentially useful in the future?  What would be your design for this table, because since CA cert differs from "regular" cert on such a low level I'm struggling to come up with a way to effectively distinguish it on DB table level.

However, the more I think about it, maybe the standalone table was bad approach after all. Maybe we should just use existing `config` table that accepts keys/values and store our CA as `ca_cert`, `ca_key` in the config database. What do you think about this approach?

fnordahl replied:
I quite like the idea of just stashing it in key/value pairs in the `config` table, go for it!

If we in the future get a need for a full blown data structure for multiple certificate types etc, we then probably also have specific use cases to feed the design work of it.

mkalcok commented on README.md:
```diff
@@ -1,3 +1,103 @@
+> **Warning**
+> 
+> Certificate and private key generated for the self-provisioned CA by MicroOVN
+> are stored unencrypted in database on every cluster member. This means that
```
:+1: added

mkalcok commented on README.md:
```diff
@@ -1,3 +1,103 @@
+> Certificate and private key generated for the self-provisioned CA by MicroOVN
+> are stored unencrypted in database on every cluster member. This means that
+> if attackers gain access to even one cluster member, they can use the CA to
+> issue valid certificates accepted by other cluster members.
```
But this will always be the case for the self-provisioned mode right? I don't think there's a way around sharing cert and private key with this approach.

mkalcok commented on README.md:
```diff
@@ -1,3 +1,103 @@
+> issue valid certificates accepted by other cluster members.
+
+For now, this self-provisioned TLS setup is the only mode available and while
+not perfect, the security it provides is vastly superior to the plaintext
```
Good point, thanks. It's reworded now.

mkalcok commented on README.md:
````diff
@@ -1,3 +1,103 @@
+microovn.ovn-appctl -t /var/snap/microovn/common/run/central/ovnnb_db.ctl cluster/status OVN_Northbound
+```
+
+## TLS encryption
````
Yup, makes sense. I added section for that

mkalcok commented on README.md:
```diff
@@ -1,3 +1,103 @@
+not perfect, the security it provides is vastly superior to the plaintext
+communication.
+
+**TODO**: In future we plan to add option for user to supply their own,
```
Yes, I scraped this whole "TODO" paragraph and instead added a sentence to the previous paragraph that is less "committing".

mkalcok commented:
Thanks for the review @fnordahl I pushed few updates and I believe I addressed all your comments

* I dropped the two superfluous commits from this branch
* Moved README changes to separate commit and applied your suggestions
* Switched to "Server" certificates for OVN controller
* Improved docstring for `Certificate` database record
* added explanation linking our choice of ECDSA+P384 with LXD.

mkalcok commented:
After discussion w/ @fnordahl I switched to using `config` database table to store CA cert and key, instead of creating dedicated `certificate` table.

fnordahl reviewed (changes requested):
Thank you for your work on this, Martin!

This is coming together nicely, and I have some comments in-line.

mkalcok commented on microovn/api/types/certificates.go:
```diff
@@ -0,0 +1,77 @@
+		)
+	}
+
+}
```
:+1: Thanks, you are right, I looked it up and the proper way is to use functions from `fmt` package. I didn't realize that `fmt` also contains `print()` and `println()` alternatives.

mkalcok commented on snapcraft.yaml:
```diff
@@ -95,6 +95,12 @@ apps:
 
+  refresh-expiring-certs:
+    command: commands/refresh-expiring-certs
+    daemon: oneshot
```
Yeah, not much we can do about naming here although it doesn't seem contradictory to me. `oneshot` simply signifies that the "service" process is expected to exit.
As for the timer and services, I don't think you are right, [the official documentation explicitly](https://www.freedesktop.org/software/systemd/man/systemd.timer.html) mentions it:

> For each timer file, a matching unit file must exist, describing the unit to activate when the timer elapses. By default, a service by the same name as the timer (except for the suffix) is activated.

mkalcok commented on snapcraft/commands/refresh-expiring-certs:
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env bash
```
I deliberated a lot about whether to implement this in `microovn` itself or as a service script and I decided to go with the script for following reasons:
* I **really** wanted to leverage existing functionality of `openssl` that just tells you if the certificate expires within specified time period or not. This saves us a lot of implementation details that we'd have to deal with if we were to implement this check ourselves.
* This is a functionality that is not useful to user in day-to-day operations. We'd be adding CLI command that would not be useful to end user, unnecessarily cluttering the command line UI.

I realize that these are not the strongest points but it's what tipped me towards the script. To your points

> without any meaningful way of testing

I'd argue that we can test this quite easily by passing argument `-t 4000` (~11 years) which would guarantee that both CA and the service certificates would be considered expiring and we could observe whether the automatic refresh process works.

>  it appears that the previous commit of adding JSON output format is also done only to provide support for the processing in this bash script.

You are right, that was the main motivation, although I'd say that having machine-readable output is never bad thing and can come in handy later even if there's no other clear use-case for it right now.

>  the addition of the jq stage-package

The `jq` package itself is very minimalistic. It has only one dependency (outside of libc) and both main package and its dependencies are just couple hundred kB in size combined. 

> The CA regenration implementation could for example also have an extra argument only allowing the current microcluster leader do the CA regeneration, avoiding race conditions.

I tried to mitigate the possible race condition by specifying time range for the daily check so that not all members execute it at the same time (we can expand the time range to further lower the race condition probability). Adding the self-imposed client-side limitation doesn't sound very good to me. It'd be again something that's not very useful to regular user and kind-of forced into the CLI for this one specific task.
I didn't test this but I suspect that the race-condition might not even be that sever. The "regenerate-ca" process has 3 steps:
* Issue new CA (distributed via dqlite)
* request peers to issue new service certificates
* issue local service certificates
Now if in the middle of this process comes another member and executes the same, my suspicion is that it'll cause a bit of thrashing but everyone will eventually end up with the CA and certificates that were generated by the member that initiated the process later. 

All this being said I'm not opposed to implementing it in `microovn` itself, as I mentioned I considered it heavily. But as you said, It'd probably push the implementation out of this pulse. Let me know if you still think that the service script is not a good idea.

fnordahl replied:
TLDR; I'm still a bit skeptic to the bash approach, but it does serve the purpose now, and we can always change it later, so will not block on it now.

> I deliberated a lot about whether to implement this in `microovn` itself or as a service script and I decided to go with the script for following reasons:
> 
>     * I **really** wanted to leverage existing functionality of `openssl` that just tells you if the certificate expires within specified time period or not. This saves us a lot of implementation details that we'd have to deal with if we were to implement this check ourselves.

The go code could still execute the `openssl` binary to do the work. My main motivation for wanting it there is just to avoid the need for exporting all this data and process it in bash.

>     * This is a functionality that is not useful to user in day-to-day operations. We'd be adding CLI command that would not be useful to end user, unnecessarily cluttering the command line UI.

That is a fair point, I do think a `--reissue-only-if-the-certificate-is-about-to-expire` (obviously has to have a different name ;)) option could be useful for the end user though.

> > without any meaningful way of testing
> 
> I'd argue that we can test this quite easily by passing argument `-t 4000` (~11 years) which would guarantee that both CA and the service certificates would be considered expiring and we could observe whether the automatic refresh process works.

If you say so, I have only had regrets when it comes to bash and testing, but I guess it's not impossible.

> > it appears that the previous commit of adding JSON output format is also done only to provide support for the processing in this bash script.
> 
> You are right, that was the main motivation, although I'd say that having machine-readable output is never bad thing and can come in handy later even if there's no other clear use-case for it right now.

Machine readable stuff is indeed good, but you're also creating an API that an end user may rely on. What if we at some point in the future do not have this information available anymore (for example because doing cert lifecycle somewhere else), would we break anyone's workflow?

> > the addition of the jq stage-package
> 
> The `jq` package itself is very minimalistic. It has only one dependency (outside of libc) and both main package and its dependencies are just couple hundred kB in size combined.

:+1:

> > The CA regenration implementation could for example also have an extra argument only allowing the current microcluster leader do the CA regeneration, avoiding race conditions.
> 
> I tried to mitigate the possible race condition by specifying time range for the daily check so that not all members execute it at the same time (we can expand the time range to further lower the race condition probability). Adding the self-imposed client-side limitation doesn't sound very good to me. It'd be again something that's not very useful to regular user and kind-of forced into the CLI for this one specific task. I didn't test this but I suspect that the race-condition might not even be that sever. The "regenerate-ca" process has 3 steps:
> 
>     * Issue new CA (distributed via dqlite)
> 
>     * request peers to issue new service certificates
> 
>     * issue local service certificates
>       Now if in the middle of this process comes another member and executes the same, my suspicion is that it'll cause a bit of thrashing but everyone will eventually end up with the CA and certificates that were generated by the member that initiated the process later.

Synchronization in clustered systems is hard, so adding CLI options to enable users to deal with it is not a bad thing imo. there is precedent for `--leader-only` / `--no-leader-only` type options in the upstream project as well, so it would be familiar. The end user would have the same problem if they wanted to manage the lifecycle themselves.

> All this being said I'm not opposed to implementing it in `microovn` itself, as I mentioned I considered it heavily. But as you said, It'd probably push the implementation out of this pulse. Let me know if you still think that the service script is not a good idea.

Yep, let's keep your current implementation and revisit in the future if needed.

mkalcok commented on snapcraft/commands/refresh-expiring-certs:
```diff
@@ -0,0 +1,76 @@
+SERVICES=$(echo "$CERT_LIST" | jq -r '. | keys[]')
+
+refresh_ca
+for s in $SERVICES; do
```
`reissue all` would indiscriminately issued new certificates every time this script ran. We want to examine each certificate, determine if it's expiring and reissue it individually only if it's within the expiration threshold.

mkalcok commented on snapcraft/commands/refresh-expiring-certs:
```diff
@@ -0,0 +1,76 @@
+EXPIRE_THRESHOLD=$(( THRESHOLD_DAYS * 24 * 60 * 60))  # days -> seconds
+SERVICES=$(echo "$CERT_LIST" | jq -r '. | keys[]')
+
+refresh_ca
```
I went into bit more detail in [this reply](https://github.com/canonical/microovn/pull/32#discussion_r1229348575) but tldr is that I tried to mitigate it by randomizing the timer.

mkalcok commented on microovn/api/certificates/regenerate_ca.go:
```diff
@@ -0,0 +1,84 @@
+		}
+
+		// Bump rest of the cluster members to reissue their certificates with new CA
+		err = cluster.Query(s.Context, true, func(ctx context.Context, c *client.Client) error {
```
The query execution eventually ends up [here](https://github.com/canonical/microcluster/blob/9b4786b0b3f0205c5a8eed4b719f02c6e4ad1a37/internal/rest/client/client.go#L203). It seems that the timeout is specified in `Context` and it it's not it'll default to `30s`

mkalcok commented on microovn/api/certificates/regenerate_ca.go:
```diff
@@ -0,0 +1,84 @@
+			result, err := microovnClient.RegenerateCA(ctx, c)
+			if err != nil {
+				clientURL := c.URL()
+				return fmt.Errorf("failed to POST to cluster member with address %q: %w", clientURL.String(), err)
```
This is a good point. Thank you. While rest of the cluster members will be contacted even in this implementation, overall result will be failure. We can handle this more gracefully.

mkalcok commented:
Thanks for your review Frode. I incorporated some of your feedback to the new version. Updates:

* Console output methods are now unified and all come from `fmt` package
* If any cluster member fails to respond during the `regenerate-ca` command, this error will be properly logged and displayed instead of whole command failing.
* `tcp`/`ssl` protocols are now configured conditionally based on whether we have a CA in database. This helps smooth out the transition from plaintext to TLS deployments. Any existing deployments that get upgraded will keep using `tcp` until manual steps are taken to enable TLS.
  * Steps required to enable TLS are included in the README.md

fnordahl reviewed (approved):
LGTM, thank you for your work on this Martin. Great stuff!
